File: apps/mobile/lib/api/models/checkout.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'converters.dart';
import 'decimal.dart';
import 'enums.dart';
import 'money.dart';
import 'order_totals.dart';

part 'checkout.freezed.dart';
part 'checkout.g.dart';

/// The ship-to address as a quote and an order carry it — exactly the four
/// required fields `/api/orders` accepts, and no more. Pricing freight against
/// a richer address than checkout admits would quote for a destination the
/// order could not be placed to.
@freezed
abstract class ShippingAddress with _$ShippingAddress {
  const factory ShippingAddress({
    required String label,
    required String line1,
    required String city,
    required Country country,
    String? line2,
    String? postalCode,
  }) = _ShippingAddress;

  factory ShippingAddress.fromJson(Map<String, dynamic> json) =>
      _$ShippingAddressFromJson(json);
}

/// One priced line of a checkout quote.
///
/// Note this line carries its own [vatAmount] while [CartLine] does not: a
/// quote is the thing an order is written from, so every figure it was built
/// out of travels with it.
@freezed
abstract class QuoteLine with _$QuoteLine {
  const QuoteLine._();

  const factory QuoteLine({
    required String productId,
    required String? variantId,
    required String sellerId,
    required String sku,
    required String nameEn,
    required String nameAr,
    required int quantity,
    @DecimalConverter() required Decimal unitPrice,
    @DecimalConverter() required Decimal vatRatePercent,
    @DecimalConverter() required Decimal vatAmount,
    @DecimalConverter() required Decimal lineTotal,
  }) = _QuoteLine;

  factory QuoteLine.fromJson(Map<String, dynamic> json) =>
      _$QuoteLineFromJson(json);

  String name(Language language) => language == Language.ar ? nameAr : nameEn;

  Money unitPriceIn(Currency currency) => Money.of(unitPrice, currency);
  Money vatAmountIn(Currency currency) => Money.of(vatAmount, currency);
  Money lineTotalIn(Currency currency) => Money.of(lineTotal, currency);
}

/// What delivery costs, or why it could not be priced.
///
/// [status] is the field to branch on. `unpriced_no_zones` is NOT an error and
/// NOT free shipping: no zone covers the address, [amount] is zero, and the
/// buyer must be told freight will be quoted separately rather than shown a
/// total that omits it.
@freezed
abstract class ShippingQuote with _$ShippingQuote {
  const ShippingQuote._();

  const factory ShippingQuote({
    required ShippingQuoteStatus status,
    required String? zoneName,
    @DecimalConverter() required Decimal amount,
    @DecimalConverter() required Decimal vatRatePercent,
    required int? estimatedDaysMin,
    required int? estimatedDaysMax,
  }) = _ShippingQuote;

  factory ShippingQuote.fromJson(Map<String, dynamic> json) =>
      _$ShippingQuoteFromJson(json);

  Money amountIn(Currency currency) => Money.of(amount, currency);

  bool get isPriced => status == ShippingQuoteStatus.priced;

  /// True when the total on screen does not include freight, because freight
  /// is not knowable yet. The checkout screen MUST say so.
  bool get freightQuotedSeparately =>
      status == ShippingQuoteStatus.unpricedNoZones;

  bool get hasEstimate => estimatedDaysMin != null && estimatedDaysMax != null;
}

@freezed
abstract class AppliedPromotion with _$AppliedPromotion {
  const AppliedPromotion._();

  const factory AppliedPromotion({
    required String promotionId,

    /// Null for an automatic promotion the buyer did not type a code for.
    required String? couponCode,
    required String label,
    @DecimalConverter() required Decimal discountAmount,
  }) = _AppliedPromotion;

  factory AppliedPromotion.fromJson(Map<String, dynamic> json) =>
      _$AppliedPromotionFromJson(json);

  Money discountIn(Currency currency) => Money.of(discountAmount, currency);
}

/// A priced basket, valid until [expiresAt].
///
/// The totals carry the goods/shipping VAT split as two separate required
/// fields, and [OrderTotals.fromJson] refuses a payload where they do not add
/// up — see `order_totals.dart` for why that check exists.
@freezed
abstract class CheckoutQuote with _$CheckoutQuote {
  const CheckoutQuote._();

  const factory CheckoutQuote({
    required String quoteId,
    required Currency currency,
    required Channel channel,

    /// The jurisdiction's headline rate, as a PERCENTAGE (5 means 5%), which
    /// is the unit `composeOrderTotals` takes. A fraction such as 0.05 here
    /// would under-tax by a factor of a hundred.
    @DecimalConverter() required Decimal vatRatePercent,
    required List<QuoteLine> lines,
    required ShippingQuote shipping,
    required List<AppliedPromotion> promotions,
    required OrderTotals totals,
    @UtcDateTimeConverter() required DateTime expiresAt,
  }) = _CheckoutQuote;

  factory CheckoutQuote.fromJson(Map<String, dynamic> json) =>
      _$CheckoutQuoteFromJson(json);

  /// Totals paired with [currency], ready to format.
  MoneyTotals get money => totals.inCurrency(currency);

  bool isExpired(DateTime now) => !now.toUtc().isBefore(expiresAt);

  Duration timeRemaining(DateTime now) {
    final left = expiresAt.difference(now.toUtc());
    return left.isNegative ? Duration.zero : left;
  }

  /// True when the quote priced VAT on the delivery as well as on the goods.
  /// A zero-rated market (QA, KW) reports false here with a zero component,
  /// which is correct and different from the component being missing.
  bool get chargesVatOnShipping => !totals.shippingVatAmount.isZero;
}

File: packages/contracts/src/checkout.ts
import { z } from "zod";

import { ChannelSchema, CountrySchema, CurrencySchema } from "./enums";
import { IdSchema, MoneySchema, TimestampSchema, VatRatePercentSchema } from "./primitives";
import { successEnvelope } from "./envelope";

/**
 * The order's ship-to address, exactly the four fields `/api/orders` requires
 * and with its bounds. A quote that priced delivery against a richer address
 * than checkout accepts would quote freight for a destination the order could
 * not be placed to.
 */
export const ShippingAddressInputSchema = z
  .object({
    label: z.string().trim().min(1).max(80),
    line1: z.string().trim().min(3).max(240),
    line2: z.string().trim().max(240).optional(),
    city: z.string().trim().min(1).max(120),
    country: CountrySchema,
    postalCode: z.string().trim().max(20).optional(),
  })
  .strict();

export const QuoteLineInputSchema = z
  .object({
    productId: IdSchema,
    variantId: IdSchema.optional(),
    quantity: z.number().int().positive().max(100_000),
  })
  .strict();

/**
 * POST /v1/checkout/quote — request.
 *
 * `items` is required and explicit even though the server holds a cart: a
 * quote the buyer is shown must be a quote for the lines the buyer can see,
 * and a server-side cart that changed between render and quote would otherwise
 * price something the screen never displayed. `cartId` travels alongside so
 * the server can refuse a quote against a cart that has moved on.
 *
 * Prices, discounts, VAT and freight are NOT accepted from the client. They
 * are resolved server-side, for the same reason `/api/orders` resolves them:
 * a shipping figure the client can influence is a discount the client can
 * grant itself.
 */
export const CheckoutQuoteRequestSchema = z
  .object({
    cartId: IdSchema.optional(),
    items: z.array(QuoteLineInputSchema).min(1).max(500),
    shippingAddress: ShippingAddressInputSchema,
    /** Required, never defaulted: a defaulted currency prices an order the buyer never chose. */
    currency: CurrencySchema,
    channel: ChannelSchema.default("B2C"),
    couponCode: z
      .string()
      .trim()
      .min(3)
      .max(40)
      .regex(/^[A-Za-z0-9_-]+$/)
      .optional(),
  })
  .strict();

export type CheckoutQuoteRequest = z.infer<typeof CheckoutQuoteRequestSchema>;

/**
 * THE ORDER TOTALS. This mirrors `OrderTotals` in
 * `packages/database/src/services/checkout-invariants.ts`, field for field.
 *
 * The seven fields are all REQUIRED and the object is `.strict()`, which is
 * the entire point of this schema. PR #21 fixed a defect where VAT was charged
 * on the goods and not on the delivery: the total was
 * `goods + goodsVat + shipping`, which adds freight AFTER tax. Every figure on
 * the order agreed with every other, the buyer was undercharged, and the
 * understated `vatAmount` was persisted for invoicing and settlement to read —
 * a silent wrong answer that scaled with the freight bill.
 *
 * A contract with a single `vatAmount` field would let that defect back in
 * without anything failing, because a collapsed VAT figure is indistinguishable
 * from a correct one at the type level. So the two components are separate,
 * neither is optional, neither is nullable, and the identities below are
 * enforced at parse time rather than documented in prose:
 *
 *   vatAmount === goodsVatAmount + shippingVatAmount
 *   total     === subtotal - discountAmount + goodsVatAmount
 *                          + shippingAmount + shippingVatAmount
 *
 * A zero-rated jurisdiction needs no special case: QA and KW carry rate 0, so
 * `shippingVatAmount` is 0 and both identities still hold. That is different
 * from the field being absent, and this schema keeps the difference.
 *
 * The comparisons are exact rather than epsilon-based on purpose. Every input
 * is already rounded to two places by the server's `money()` helper before it
 * is summed, so the sum of the rounded parts is exact in binary floating point
 * at these magnitudes; an epsilon here would forgive precisely the drift the
 * schema exists to catch. The one place rounding could bite — the final
 * `total`, rounded once from already-rounded parts — is compared through the
 * same `toFixed(2)` the server applies.
 */
const roundMoney = (value: number) => Number(value.toFixed(2));

export const OrderTotalsSchema = z
  .object({
    /** Goods, net of VAT, before discount. */
    subtotal: MoneySchema,
    discountAmount: MoneySchema,
    /** VAT on the goods, after discount. */
    goodsVatAmount: MoneySchema,
    /** Delivery, net of VAT. */
    shippingAmount: MoneySchema,
    /** VAT on the delivery, at the order's place-of-supply rate. */
    shippingVatAmount: MoneySchema,
    /** What the invoice declares as tax: goods VAT plus delivery VAT. */
    vatAmount: MoneySchema,
    total: MoneySchema,
  })
  .strict()
  .superRefine((totals, ctx) => {
    if (roundMoney(totals.goodsVatAmount + totals.shippingVatAmount) !== totals.vatAmount) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ["vatAmount"],
        message: "vatAmount must equal goodsVatAmount + shippingVatAmount — declared tax is both components, never only the goods",
      });
    }
    const expectedTotal = roundMoney(
      totals.subtotal - totals.discountAmount + totals.goodsVatAmount + totals.shippingAmount + totals.shippingVatAmount,
    );
    if (expectedTotal !== totals.total) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ["total"],
        message: "total must equal subtotal - discountAmount + goodsVatAmount + shippingAmount + shippingVatAmount",
      });
    }
  });

export type OrderTotals = z.infer<typeof OrderTotalsSchema>;

/** The goods half of an order — what a governed PO snapshot approves. */
export const merchandiseTotalOf = (totals: OrderTotals): number =>
  roundMoney(totals.subtotal - totals.discountAmount + totals.goodsVatAmount);

/**
 * One priced line of the quote. `unitPrice` is the tier the server resolved
 * for THIS quantity, which is why the app must not reprice a tiered line
 * locally — see `priceTiered` on the cart line.
 */
export const QuoteLineSchema = z
  .object({
    productId: IdSchema,
    variantId: IdSchema.nullable(),
    sellerId: IdSchema,
    sku: z.string().min(1).max(64),
    nameEn: z.string().min(1).max(200),
    nameAr: z.string().min(1).max(200),
    quantity: z.number().int().positive().max(100_000),
    unitPrice: MoneySchema,
    /** Per-line VAT rate as a percentage. A seller may configure this per product. */
    vatRatePercent: VatRatePercentSchema,
    vatAmount: MoneySchema,
    lineTotal: MoneySchema,
  })
  .strict();

/**
 * How delivery was priced.
 *
 * `status` distinguishes the three real answers `secureCreateOrder` already
 * gives, which a bare `shippingAmount: 0` cannot: no zones are configured at
 * all (freight is genuinely zero and the order proceeds), a zone covers the
 * destination (a real price), or zones exist and none covers it (the order
 * will be REFUSED). An app that cannot tell the first from the third shows a
 * free-delivery badge on an order that is about to be rejected.
 */
export const ShippingQuoteSchema = z
  .object({
    status: z.enum(["priced", "unpriced_no_zones", "unavailable"]),
    zoneName: z.string().max(120).nullable(),
    amount: MoneySchema,
    /** Place-of-supply rate applied to the delivery. Percentage: 5 means 5%. */
    vatRatePercent: VatRatePercentSchema,
    estimatedDaysMin: z.number().int().min(0).max(365).nullable(),
    estimatedDaysMax: z.number().int().min(0).max(365).nullable(),
  })
  .strict();

export const AppliedPromotionSchema = z
  .object({
    promotionId: IdSchema,
    couponCode: z.string().max(40).nullable(),
    label: z.string().max(200),
    discountAmount: MoneySchema,
  })
  .strict();

/**
 * POST /v1/checkout/quote — response payload.
 *
 * `vatRatePercent` is the ORDER's place-of-supply rate, the one the delivery
 * is taxed at. It is stated separately from the per-line rates because the
 * delivery is a supply by the platform to the buyer and follows the order's
 * place of supply, not whatever rate a seller configured on a product.
 */
export const CheckoutQuoteSchema = z
  .object({
    quoteId: IdSchema,
    currency: CurrencySchema,
    channel: ChannelSchema,
    /** The order's place-of-supply VAT rate, as a percentage (5 means 5%). */
    vatRatePercent: VatRatePercentSchema,
    lines: z.array(QuoteLineSchema).min(1).max(500),
    shipping: ShippingQuoteSchema,
    promotions: z.array(AppliedPromotionSchema),
    totals: OrderTotalsSchema,
    /**
     * When this quote stops being the price. Prices, stock and promotions all
     * move; a quote with no expiry invites the app to show a stale total.
     */
    expiresAt: TimestampSchema,
  })
  .strict();

export type CheckoutQuote = z.infer<typeof CheckoutQuoteSchema>;

export const CheckoutQuoteResponseSchema = successEnvelope(CheckoutQuoteSchema);
